### ISSUE TITLE
Explicit frequency domain operations in gravmag.transform

### ISSUE THREAD

**Issue by leouieda** [fatiando/fatiando#237]
The functions that operate in the frequency domain in `fatiando.gravmag.transform` all do so implicitly. They only take arguments in the space domain and only return after inverse FFT is applied. As pointed out by @santis19 in #230, this can be a problem when chaining multiple operations in large datasets. 

I can think of the following changes that could solve this problem:

1. Make two sets of functions for each transformation: one that takes space domain input and return space domain output (these will be the ones that we already have); another that takes input in the frequency domain and returns in the frequency domain. The space domain functions can use the frequency domain ones internally so code is not duplicated and the user has a choice.
2. Make the FFT utility functions public (better arguments, names, docstrings, tests). These will be needed for the space domain functions and the user will need them when using the frequency domain functions directly. 

One problem that we might run into is the issue of padding the arrays before FFT. That should be a utility function as well and the frequency domain functions should recommend their usage.

Relevant to #225 as well.

**Comment by drandykass** [comment]
Agreed that some of the fft utility functions should be public.  I have
written prism_fourier to expect padded and gridded data locations, but it
would be nice to have the routines pad as necessary.

On Wed, Nov 25, 2015 at 2:06 PM, Leonardo Uieda <notifications@github.com>
wrote:

> The functions that operate in the frequency domain in
> fatiando.gravmag.transform all do so implicitly. They only take arguments
> in the space domain and only return after inverse FFT is applied. As
> pointed out by @santis19 <https://github.com/santis19> in #230
> <https://github.com/fatiando/fatiando/pull/230>, this can be a problem
> when chaining multiple operations in large datasets.
>
> I can think of the following changes that could solve this problem:
>
>    1. Make two sets of functions for each transformation: one that takes
>    space domain input and return space domain output (these will be the ones
>    that we already have); another that takes input in the frequency domain and
>    returns in the frequency domain. The space domain functions can use the
>    frequency domain ones internally so code is not duplicated and the user has
>    a choice.
>    2. Make the FFT utility functions public (better arguments, names,
>    docstrings, tests). These will be needed for the space domain functions and
>    the user will need them when using the frequency domain functions directly.
>
> One problem that we might run into is the issue of padding the arrays
> before FFT. That should be a utility function as well and the frequency
> domain functions should recommend their usage.
>
> Relevant to #225 <https://github.com/fatiando/fatiando/issues/225> as
> well.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fatiando/fatiando/issues/237>.
>



-- 
------------------------------------------------------------------------------------
Dum inter homines sumus, colamus humanitatem.
------------------------------------------------------------------------------------

**Comment by leouieda** [comment]
The problem is that padding needs to be done before FFT, right?

**Comment by drandykass** [comment]
Yes, but the prism Fourier is computed directly in the Fourier domain, and
just needs to know dx/dy and nx/ny. So really you just need to generate a
padded xp and yp before starting the modelling.

**Comment by drandykass** [comment]
But having said that, it's no worries for now to simply require the input
be gridded and a power of 2 in each dimension.

On Wed, Nov 25, 2015 at 3:22 PM, M. Andy Kass <kass.andrew@gmail.com> wrote:

> Yes, but the prism Fourier is computed directly in the Fourier domain, and
> just needs to know dx/dy and nx/ny. So really you just need to generate a
> padded xp and yp before starting the modelling.
>
> Sent from mobile device. Please excuse errors.
> On Nov 25, 2015 2:37 PM, "Leonardo Uieda" <notifications@github.com>
> wrote:
>
>> The problem is that padding needs to be done before FFT, right?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/fatiando/fatiando/issues/237#issuecomment-159736731>.
>>
>


-- 
------------------------------------------------------------------------------------
Dum inter homines sumus, colamus humanitatem.
------------------------------------------------------------------------------------

**Comment by leouieda** [comment]
> Yes, but the prism Fourier is computed directly in the Fourier domain, and just needs to know dx/dy and nx/ny. So really you just need to generate a padded xp and yp before starting the modelling.

Agree. That's why a `pad_arrays` functions would be really useful. Maybe in the `gridder` module? It could also be used by the functions in `gravmag.transform`.

**Comment by drandykass** [comment]
I like it.

Do you think you send pad_arrays the same inputs as gridder.regular, and it
returns a padded xp,yp,zp and additionally padx, pady (calling
gridder.regular internally)?  Or would it be better to call
gridder.regular, then send gridder.pad_arrays the output from
gridder.regular?

This may be a great thing for me to do really quickly and practice
submitting a pull request--simpler than the big pull request adding all the
fourier modelling business.  Thoughts?

**Comment by leouieda** [comment]
>Do you think you send pad_arrays the same inputs as gridder.regular, and it returns a padded xp,yp,zp and additionally padx, pady (calling gridder.regular internally)? Or would it be better to call gridder.regular, then send gridder.pad_arrays the output from gridder.regular?

I think it's better to have it independent. That way it's also useful if you already have gridded data somehow. So it could just take the x, y, data arrays and apply the padding. It should return the padded arrays as well as the new shape of the arrays. It would probably be useful to have a `unpad_arrays` then to undo the process.

> This may be a great thing for me to do really quickly and practice submitting a pull request--simpler than the big pull request adding all the fourier modelling business. Thoughts? 

Completely agree! This is a great test project. You should make these changes in a separate branch (based on the `master` branch) and submit a separate PR. This way we can merge this independently of the other contributions.

**Comment by drandykass** [comment]
Consider it done!

**Comment by leouieda** [comment]
:+1:

**Comment by drandykass** [comment]
OK, I have written a padding routine which will take an array of arbitrary
dimension and pad it to either a user supplied set of values or to the next
power of 2 if none.  It will pad using a constant, the edge value, the mean
value, even or odd reflection, linear taper to the mean, or odd reflection
with a cosine taper (the best, and the biggest pain to code up!), depending
on user input.  It returns a deep copy of the padded array (or rather a
deep copy of the original array put in the padded array), plus a padded
list of coordinate vectors.

I need to clean it up and write some test programs (in Fatiando format
rather than my own unittests), then I can make a pull request.  Or, I can
do as you suggested and make the pull request now before it's complete.
Just let me know.

On Fri, Nov 27, 2015 at 3:30 AM, Leonardo Uieda <notifications@github.com>
wrote:

> [image: :+1:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fatiando/fatiando/issues/237#issuecomment-160107861>.
>



-- 
------------------------------------------------------------------------------------
Dum inter homines sumus, colamus humanitatem.
------------------------------------------------------------------------------------

**Comment by drandykass** [comment]
An example of the possible outputs (for a 2-D array. These will natively work for a 1D or arbitrary dimension array too):

![padding_algorithms](https://cloud.githubusercontent.com/assets/15897239/11512320/ca8f403e-982a-11e5-8bd0-36ce0bcc8d6c.png)

**Comment by leouieda** [comment]
@drandykass wow that's way more than I expected! Great! 

You can open the PR whenever you want. No need to wait for things to be finished. That way TravisCI will pick it up and start running the tests so we can check if all works.  

Could you put the code for this figure as a cookbook recipe as well?

**Comment by drandykass** [comment]
Yes, should I add a file into the cookbook as part of this branch?

On Wed, Dec 2, 2015 at 12:58 PM, Leonardo Uieda <notifications@github.com>
wrote:

> @drandykass <https://github.com/drandykass> wow that's way more than I
> expected! Great!
>
> You can open the PR whenever you want. No need to wait for things to be
> finished. That way TravisCI will pick it up and start running the tests so
> we can check if all works.
>
> Could you put the code for this figure as a cookbook recipe as well?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fatiando/fatiando/issues/237#issuecomment-161415339>.
>



-- 
------------------------------------------------------------------------------------
Dum inter homines sumus, colamus humanitatem.
------------------------------------------------------------------------------------

**Comment by leouieda** [comment]
Yes, that would be great!
